### PR TITLE
Support fulfill order with 0 quantity only if total quantity is larger than 0 in dashboard and API

### PR DESCRIPTION
**Pull request by jxltom**
If an order has purchased 1 product A and 1 product B, it makes sense to only fulfill 1 product A or only 1 product B. This PR fixes this issue in dashboard as well as in API

Update: closes https://github.com/mirumee/saleor/issues/3758

### Screenshots

Before:
screenshot from 2019-02-20 14-23-02

After:
screenshot from 2019-02-20 17-18-28

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

File: saleor/order/models.py [hunk 1]
```diff
@@ -315,6 +315,8 @@ class Meta:
     def __str__(self):
         return self.product_name
 
+    __repr__ = __str__
```
Comment (maarcingebala) [inline saleor/order/models.py]: This change seems unrelated, I'd remove it.

File: saleor/dashboard/order/forms.py [hunk 2]
```diff
@@ -614,6 +614,16 @@ def __init__(self, *args, **kwargs):
         for form in self.forms:
             form.empty_permitted = False
 
+    def clean(self):
+        if any(self.errors):
```
Comment (maarcingebala) [inline saleor/dashboard/order/forms.py]: Why do we need that check? It's rather uncommon in clean methods and in our codebase as well I think. I've tested this code and if we removed it and below used `form.cleaned_data.get('quantity', 0)`, all errors would be returned properly.

File: saleor/dashboard/order/forms.py [hunk 3]
```diff
@@ -614,6 +614,16 @@ def __init__(self, *args, **kwargs):
...
+            return
+
+        total_quantities = sum(
+            form.cleaned_data['quantity'] for form in self.forms)
+        if total_quantities <= 0:
```
Comment (maarcingebala) [inline saleor/dashboard/order/forms.py]: I'm not 100% if we need this validation. If we didn't have it, fulfilling 0 quantities for all lines would redirect the user to order details view with a message "No lines fulfilled" which quite makes sense.

File: saleor/dashboard/order/views.py [hunk 4]
```diff
@@ -591,7 +591,7 @@ def fulfill_order_lines(request, order_pk):
         request.POST or None, queryset=FulfillmentLine.objects.none(),
         initial=initial)
     all_forms_valid = all([line_form.is_valid() for line_form in formset])
-    if all_forms_valid and form.is_valid():
+    if all_forms_valid and formset.is_valid() and form.is_valid():
```
Comment (maarcingebala) [inline saleor/dashboard/order/views.py]: Maybe we should rename `all_forms_valid` to sth like `are_line_forms_valid` - now this naming suggests that literally all forms used here are valid while in the `if` check we still check `formset` and `form`.

File: saleor/order/models.py [hunk 5]
```diff
@@ -367,7 +369,7 @@ class FulfillmentLine(models.Model):
         OrderLine, related_name='+', on_delete=models.CASCADE)
     fulfillment = models.ForeignKey(
         Fulfillment, related_name='lines', on_delete=models.CASCADE)
-    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(0)])
```
Comment (maarcingebala) [inline saleor/order/models.py]: With this change, this field basically becomes a `PositiveIntegerField`.

File: saleor/dashboard/order/forms.py (hunk 2 above)
Comment (jxltom) [reply saleor/dashboard/order/forms.py]: It's just for bypassing all following validations in advance if there are errors already, which makes the clean method faster. I will remove it if it's not suitable for saleor codebase.

File: saleor/dashboard/order/forms.py (hunk 3 above)
Comment (jxltom) [reply saleor/dashboard/order/forms.py]: But that gives wrong impression to users that it looks like the fulfillment is succeeded. Actually it has errors and the current form should reflect the errors. The user would like to fulfill the order indeed and so we shouldn't redirect to other pages if the form has errors.

File: saleor/dashboard/order/views.py (hunk 4 above)
Comment (jxltom) [reply saleor/dashboard/order/views.py]: Will do. :D

File: saleor/order/models.py (hunk 5 above)
Comment (jxltom) [reply saleor/order/models.py]: Sounds more cleaner.

File: saleor/dashboard/order/forms.py (hunk 3 above)
Comment (maarcingebala) [reply saleor/dashboard/order/forms.py]: OK, let's leave this validation then 👍

**Comment by jxltom** [comment]
@maarcingebala The PR is revised accordingly.
